pkg/controller/standard.oam.dev/v1alpha1/rollout/handler.go

```diff
@@ -0,0 +1,225 @@
+limitations under the License.
+*/
+
+package rollout
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/oam-dev/kubevela/pkg/oam"
+
+	"github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationrollout"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"k8s.io/klog/v2"
+
+	"github.com/oam-dev/kubevela/apis/standard.oam.dev/v1alpha1"
+
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/application/assemble"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+type handler struct {
+	*reconciler
+	rollout       *v1alpha1.Rollout
+	sourceRevName string
+	targetRevName string
+	compName      string
+
+	sourceWorkload *unstructured.Unstructured
+	targetWorkload *unstructured.Unstructured
+}
+
+// assembleWorkload help to assemble worklad info, because rollout don't know AppRevision we cannot use assemble func directly.
+// but we do same thing with it in the func
+func (h *handler) assembleWorkload(ctx context.Context) error {
+	workloadOptions := []assemble.WorkloadOption{
+		applicationrollout.RolloutWorkloadName(h.compName),
+		assemble.PrepareWorkloadForRollout(h.compName),
+		applicationrollout.HandleReplicas(ctx, h.compName, h)}
+
+	for _, workloadOption := range workloadOptions {
+		if err := workloadOption.ApplyToWorkload(h.targetWorkload, nil, nil); err != nil {
+			return err
+		}
+		if len(h.sourceRevName) != 0 {
+			if err := workloadOption.ApplyToWorkload(h.sourceWorkload, nil, nil); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// extractWorkloadFromCompRevision extract workload info from component revision
+func (h *handler) extractWorkloadFromCompRevision(ctx context.Context) error {
+	var err error
+	var targetRevsion v1.ControllerRevision
+	if err = h.Get(ctx, types.NamespacedName{Namespace: h.rollout.Namespace, Name: h.targetRevName}, &targetRevsion); err != nil {
+		return err
+	}
+	// extract component data from controllerRevision
+	targetComp, err := util.RawExtension2Component(targetRevsion.Data)
+	if err != nil {
+		return err
+	}
+	h.targetWorkload, err = util.RawExtension2Unstructured(&targetComp.Spec.Workload)
+	if err != nil {
+		return err
+	}
+	if len(h.sourceRevName) != 0 {
+		var sourceRevision v1.ControllerRevision
+		if err = h.Get(ctx, types.NamespacedName{Namespace: h.rollout.Namespace, Name: h.sourceRevName}, &sourceRevision); err != nil {
+			return err
+		}
+		sourceComp, err := util.RawExtension2Component(sourceRevision.Data)
+		if err != nil {
+			return err
+		}
+		h.sourceWorkload, err = util.RawExtension2Unstructured(&sourceComp.Spec.Workload)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// templateTargetWorkload check the target workload whether exist. if not create it.
+func (h *handler) templateTargetWorkload(ctx context.Context) error {
```

Comment:
why you call it template while you're apply it?

Reply:
just name it with appRollout's template phase. have rename it with applyTargetWorkload